builder/proxmox/common/step_start_vm.go

```diff
@@ -277,33 +279,94 @@ func generateProxmoxNetworkAdapters(nics []NICConfig) proxmox.QemuDevices {
 	}
 	return devs
 }
-func generateProxmoxDisks(disks []diskConfig) proxmox.QemuDevices {
-	devs := make(proxmox.QemuDevices)
-	for idx := range disks {
-		devs[idx] = make(proxmox.QemuDevice)
-		setDeviceParamIfDefined(devs[idx], "type", disks[idx].Type)
-		setDeviceParamIfDefined(devs[idx], "size", disks[idx].Size)
-		setDeviceParamIfDefined(devs[idx], "storage", disks[idx].StoragePool)
-		setDeviceParamIfDefined(devs[idx], "cache", disks[idx].CacheMode)
-		setDeviceParamIfDefined(devs[idx], "format", disks[idx].DiskFormat)
-
-		if (devs[idx]["type"] == "scsi" || devs[idx]["type"] == "virtio") &&
-			disks[idx].IOThread {
-			devs[idx]["iothread"] = "true"
-		}
 
-		if disks[idx].Discard {
-			devs[idx]["discard"] = "on"
-		} else {
-			devs[idx]["discard"] = "ignore"
+func generateProxmoxDisks(disks []diskConfig) *proxmox.QemuStorages {
+	ideDisks := proxmox.QemuIdeDisks{}
+	sataDisks := proxmox.QemuSataDisks{}
+	scsiDisks := proxmox.QemuScsiDisks{}
+	virtIODisks := proxmox.QemuVirtIODisks{}
+
+	ideCount := 0
+	sataCount := 0
+	scsiCount := 0
+	virtIOCount := 0
```

Comment:
I wonder what are those counts used for at this step? I understand why we'd need them during validate in order to enforce the maximum number for each type, but here it seems they're incremented but not actually used?

Reply:
These are used further down at #320, 321, 334, 335, 347, 348, 360 and 361 to iterate each disk device being passed in for mapping.

The upstream proxmox API now requires each type of disk to be in a disk bus specific struct as opposed to the previous generic QemuDevice struct. Using SATA disks as an example, each new SATA disk is created as a single QemuSataStorage struct, then mapped to a specific slot in a QemuSataDisks struct containing all the ide disks to be mapped to a VM. The [QemuSataDisks](https://pkg.go.dev/github.com/Telmate/proxmox-api-go/proxmox#QemuSataDisks) struct has disk slots hard coded into the struct so each QemuSataStorage disk as they are created needs to be assigned to a slot. The counts are used to iteratively assign a disk to a bus slot in the order they're presented to the function (basically in the order they're defined .packer.hcl or json config). 

I implemented it this way only because it aligns with the behaviour of the old generateProxmoxDisks code which returned an array of QemuDevices in the order they were presented to the function. My main goal here was to avoid introducing breaking changes for disk configuration :wink:

Reply:
Ah understood, I missed those, good call!

Reply:
To be fair, I don't think you needed to preserve exactly how the function works, it's not exposed to users directly and is private to that package, so if an updated signature makes sense, feel free to remodel it as you see fit!
EDIT: I see it's called directly as the `Disks` attribute for the qemu config structure, then I assume this is the order in which it will be exposed by qemu at the end of the day, so it may make sense to maintain the current semantics indeed.